File: balanced_brackets.py
#!/usr/bin/env python
# http://rosettacode.org/wiki/Balanced_brackets

def balanced(s):
    left_count = 0
    right_count = 0
    for i in xrange(0,len(s)):
        if s[i] == '[':
            left_count += 1
        elif s[i] == ']':
            right_count += 1
        if left_count < right_count:
            return False
    return left_count == right_count


none = "" # True
one = "[]" # True
two = "][" # False
three = "[][]" # True
four = "][][" # False
five = "[[][]]" # True
six = "[]][[]" # False

for s in [none,one,two,three,four,five,six]:
    print(s, balanced(s))

File: balanced_brackets.scala
#!/usr/bin/env scala

def balancedBrackets(s: String, i: Int = 0, r: Int = 0, l: Int = 0): Boolean = {
  if (r < l) {
    false
  } else if (i == s.length) {
    r == l
  } else if (s(i).toString == "[") {
    balancedBrackets(s, i + 1, r + 1, l)
  } else if (s(i).toString == "]") {
    balancedBrackets(s, i + 1, r, l + 1)
  } else {
    balancedBrackets(s, i + 1, r, l)
  }
}

val none = "" // True
val one = "[]" // True
val two = "][" // False
val three = "[][]" // True
val four = "][][" // False
val five = "[[][]]" // True
val six = "[]][[]" // False

println(balancedBrackets(none))
println(balancedBrackets(one))
println(balancedBrackets(two))
println(balancedBrackets(three))
println(balancedBrackets(four))
println(balancedBrackets(five))
println(balancedBrackets(six))

